### PR TITLE
Limit tags in custom study dialog to those in the selected deck

### PR DESCRIPTION
**Pull request by bogdanteleaga**
Pretty much a port of https://github.com/ankidroid/Anki-Android/pull/3857

### CODE REVIEW

File: aqt/taglimit.py [hunk 1]
```diff
@@ -19,7 +19,7 @@ def __init__(self, mw, parent):
         self.exec_()
 
     def rebuildTagList(self):
-        usertags = self.mw.col.tags.all()
+        usertags = self.mw.col.decks.getTags(self.deck['id'], False)
```
Comment (hssm) [inline aqt/taglimit.py]: The second parameter should be `True`.

**Comment by hssm** [comment]
I would suggest moving this code to the tag manager instead (`tags.py`) since it doesn't really have much to do with managing decks. It also means you don't have to duplicate the split function.

A good name for the function would be `byDeck` which follows the naming convention used in other parts of the code (e.g., the deck manager has `byName`). It'll be clear what you are asking for when you do `col.tags.byDeck(did, children=True)`.

**Comment by dae** [comment]
I agree it's more useful to show only relevant tags, but don't like that this is moving away from the tag cache - on a slow disk with 100k+ notes in a deck there may be a decent delay before the dialog is shown. For the next major update I'm considering moving back to a 1 deck per file model like Anki 1 had, which will indirectly solve this problem and also get rid of the issues people have with their tags list getting messy when they add and remove a bunch of shared decks.

**Comment by julienbaley** [comment]
Going back to one deck per file? Please involve the community a bit more before making that decision. I personally think it's a terrible idea (why not drop the concept of notes while we're at it..): seamlessly renaming decks, having sub-decks, having 1 point of entry to all my Anki data, those are extremely important to me and were the major improvements of Anki 2.

I think the last months' pull requests have shown that there is an interest from the community to be more involved, please don't work in secret. I, for one, would like to be able to write tools to complement my use of Anki, and having no idea what the API is going to be tomorrow is really unpleasant.

File: aqt/taglimit.py (hunk 1 above)
Comment (bogdanteleaga) [inline aqt/taglimit.py]: Whoops, fixed

**Comment by bogdanteleaga** [comment]
> I would suggest moving this code to the tag manager instead 

Both make sense to me, but I implemented it your way in the end.

> I agree it's more useful to show only relevant tags, but don't like that this is moving away from the tag cache - on a slow disk with 100k+ notes in a deck there may be a decent delay before the dialog is shown.

I don't imagine this is a extremely common operation, so a bit of delay might be acceptable? If it's really needed some caching could be implemented at a later point.

**Comment by timrae** [comment]
> I don't imagine this is a extremely common operation, so a bit of delay might be acceptable? If it's really needed some caching could be implemented at a later point.

I just gave it a go on my phone in AnkiDroid with my main Japanese deck (~ 25,000 notes) and there was a pretty noticeable delay on the order of 2s. Smaller decks were pretty much instantaneous. Building a cache for each deck sounds like a good idea

**Comment by timrae** [comment]
For what it's worth, after just seeing this in action on AnkiDroid I feel this will be an extremely valuable change and strongly support including it in an Anki point release.

**Comment by julienbaley** [comment]
@timrae is your phone old? I've just tried with my 17000 card deck and haven't seen any sort of delay.

> I don't imagine this is a extremely common operation, so a bit of delay might be acceptable?
I would guess so. I mean, who would complain about waiting two seconds for a rare action to complet when they have 100k+ cards in a deck?
And wouldn't it be possible to run the query in the background when a person opens the "custom study" menu? By the time the user would have ticked the box and press "choose tags", the list of tags would be established.

**Comment by timrae** [comment]
@julienbaley 

> is your phone old? I've just tried with my 17000 card deck and haven't seen any sort of delay.

It's an Xperia XR, which isn't that old, but definitely not a modern flagship device. I'd agree that a few seconds delay would be acceptable to most users with decks that big; you'd just do it in an AsyncTask and show a progress bar. In saying that though, there's already a cache for the tags in the db (col.tags), and if @dae is concerned about the performance it should be relatively easy to add a second per-deck cache next door to it or something.

**Comment by bogdanteleaga** [comment]
I'd assume this would be more of a performance hit on older PC's with huge decks.

I'm certainly no expert in databases but could this be optimized by just building some other index?

As for the other 2 options, building a per deck tag cache would probably be more transparent across implementations compared to executing asynchronously. However I'm afraid I'm probably out of my depth here with both so if someone else wants to take any of these on that'd be amazing.

Fwiw, I didn't notice any lag either on a 10k cards deck using a Oneplus One that I just imported now.

**Comment by hssm** [comment]
It's likely that you lose more time sifting through all your inapplicable tags than the upfront computation to remove them. The former is much more tedious, for sure.

**Comment by dae** [comment]
Yep, you make a good point. Thanks for the feedback and testing everyone, and I'll merge this in the next few days.

Julien: once I have some concrete ideas I will create some threads on the forum, like we did with the major changes in Anki 2.

**Pull request merged**